File: frontend/src/api/industrialSafety.ts
import { apiClient } from "@/api/client";

/**
 * Контур ПромБеза (Доп. №1 разд. 54.2): реестр опасных производственных
 * объектов.
 *
 * Ручки гейтятся модулем `industrial_safety`: у арендатора без выдачи — 404, у
 * выдававшегося-отключённого чтение остаётся (read-only, BIZ-61).
 */

export type OpoHazardClass = "I" | "II" | "III" | "IV";

/**
 * Подписи классов — словами. Сервер отдаёт готовую подпись в `hazard_class_label`,
 * это запас для мест, где на руках только код (например, разрез сводки).
 */
export const OPO_HAZARD_CLASS_TITLES: Record<string, string> = {
  I: "I класс — чрезвычайно высокая опасность",
  II: "II класс — высокая опасность",
  III: "III класс — средняя опасность",
  IV: "IV класс — низкая опасность",
};

export type HazardousFacilityDto = {
  id: string;
  name: string;
  register_number: string;
  hazard_class: string;
  hazard_class_label: string;
  site_id?: string | null;
  registered_on?: string | null;
  excluded_on?: string | null;
  status: string;
  status_label: string;
  responsible?: string | null;
  notes?: string | null;
};

export type IndustrialReadinessDto = {
  total_facilities: number;
  /** Класс → число ДЕЙСТВУЮЩИХ объектов; ключи всегда все четыре. */
  by_class: Record<string, number>;
  excluded_facilities: number;
  /** Разд. 54.2: считаются только эксплуатируемые устройства. */
  total_devices: number;
  epb_overdue: number;
  epb_due_soon: number;
  /**
   * ФАКТ: назначенный срок службы истёк, действующего заключения ЭПБ нет.
   * ГРАНИЦА: платформа НЕ решает, обязана ли экспертиза быть проведена —
   * это зависит от типа устройства, документации и норм ФНП.
   */
  devices_past_lifetime_without_epb: number;
  /** Разд. 54.2 «история работ»: устройства без единой записи о работах. */
  devices_without_work_record: number;
  /** Разд. 54.2 «аттестация»: считаются только записи с областью из справочника. */
  attestations_total: number;
  attestations_overdue: number;
  attestations_due_soon: number;
  /**
   * Разд. 54.2 «производственный контроль». ГРАНИЦА: это ФАКТ наличия плана на
   * текущий год, а не приговор — обязанность вести ПК зависит от того,
   * эксплуатирует ли организация ОПО.
   */
  current_year_plan_exists: boolean;
  pc_measures_overdue: number;
  pc_measures_planned: number;
  /** Доп. №1 разд. 57.4: открытые происшествия этой дисциплины (срез-49). */
  incidents_open: number;
};

export type PcPlanDto = {
  id: string;
  year: number;
  title: string;
  responsible?: string | null;
  approved_on?: string | null;
  status: string;
  status_label: string;
  notes?: string | null;
  measures_total: number;
  measures_overdue: number;
};

export type PcMeasureDto = {
  id: string;
  plan_id: string;
  section: string;
  section_label: string;
  title: string;
  due_on: string;
  responsible?: string | null;
  /** planned | overdue | done | cancelled — «просрочено» считает сервер. */
  status: string;
  status_label: string;
  completed_on?: string | null;
  result?: string | null;
};

export type OpoAttestationDto = {
  id: string;
  person_id: string;
  person_name: string;
  name: string;
  area_code: string;
  area_label: string;
  issued_at?: string | null;
  expires_at?: string | null;
  /** ok | due_soon | overdue | absent — «срок не указан» отдельно. */
  validity_status: string;
  validity_status_label: string;
};

/** Подписи результатов работ — запас; перевод делает сервер. */
export const OPO_WORK_RESULT_TITLES: Record<string, string> = {
  passed: "Пригодно к эксплуатации",
  with_remarks: "Пригодно с условиями",
  failed: "Не пригодно",
};

export type DeviceWorkDto = {
  id: string;
  device_id: string;
  kind: string;
  kind_label: string;
  performed_on: string;
  result: string;
  result_label: string;
  performer?: string | null;
  conclusion_number?: string | null;
  notes?: string | null;
  next_due?: string | null;
  /** Перенесла ли работа срок эксплуатации; переносит только положительная ЭПБ. */
  shifted_due: boolean;
};

export type TechnicalDeviceDto = {
  id: string;
  facility_id: string;
  kind: string;
  kind_label: string;
  name: string;
  serial_number?: string | null;
  commissioned_on?: string | null;
  lifetime_until?: string | null;
  epb_conclusion_number?: string | null;
  epb_registered_on?: string | null;
  epb_valid_until?: string | null;
  status: string;
  status_label: string;
  notes?: string | null;
  /** ok | due_soon | overdue | absent — «нет заключения» отдельное состояние. */
  epb_status: string;
  epb_status_label: string;
  past_lifetime: boolean;
  /** Последняя подтверждённая работа: срок без неё — обещание, не доказательство. */
  last_work_on?: string | null;
  last_work_result?: string | null;
};

export const industrialSafetyApi = {
  listFacilities: async (): Promise<HazardousFacilityDto[]> => {
    const { data } = await apiClient.get<{ items?: HazardousFacilityDto[] }>(
      "/industrial-safety/facilities",
      { params: { limit: 200, offset: 0 } },
    );
    return Array.isArray(data?.items) ? data.items : [];
  },

  listDevices: async (): Promise<TechnicalDeviceDto[]> => {
    const { data } = await apiClient.get<{ items?: TechnicalDeviceDto[] }>(
      "/industrial-safety/devices",
      { params: { limit: 200, offset: 0 } },
    );
    return Array.isArray(data?.items) ? data.items : [];
  },

  listAttestations: async (): Promise<OpoAttestationDto[]> => {
    const { data } = await apiClient.get<{ items?: OpoAttestationDto[] }>(
      "/industrial-safety/attestations",
      { params: { limit: 200, offset: 0 } },
    );
    return Array.isArray(data?.items) ? data.items : [];
  },

  listPcPlans: async (): Promise<PcPlanDto[]> => {
    const { data } = await apiClient.get<{ items?: PcPlanDto[] }>(
      "/industrial-safety/pc-plans",
      { params: { limit: 50, offset: 0 } },
    );
    return Array.isArray(data?.items) ? data.items : [];
  },

  listPcMeasures: async (): Promise<PcMeasureDto[]> => {
    const { data } = await apiClient.get<{ items?: PcMeasureDto[] }>(
      "/industrial-safety/pc-measures",
      { params: { limit: 200, offset: 0 } },
    );
    return Array.isArray(data?.items) ? data.items : [];
  },

  readiness: async (): Promise<IndustrialReadinessDto> => {
    const { data } = await apiClient.get<IndustrialReadinessDto>(
      "/industrial-safety/readiness",
    );
    return data;
  },
};

File: backend/app/schemas/industrial_safety.py
"""Схемы контура ПромБеза (Доп. №1 разд. 54.2): реестр ОПО."""

from __future__ import annotations

from datetime import date

from pydantic import Field

from app.schemas.base import BaseSchema


class HazardousFacilityCreate(BaseSchema):
    name: str = Field(min_length=1, max_length=255)
    #: обязателен: ОПО без номера в госреестре не существует
    register_number: str = Field(min_length=1, max_length=64)
    hazard_class: str = Field(min_length=1, max_length=8)
    site_id: str | None = Field(default=None, min_length=1, max_length=36)
    registered_on: date | None = None
    excluded_on: date | None = None
    status: str = Field(default="registered", max_length=16)
    responsible: str | None = Field(default=None, max_length=255)
    notes: str | None = None


class HazardousFacilityUpdate(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    register_number: str | None = Field(default=None, min_length=1, max_length=64)
    hazard_class: str | None = Field(default=None, min_length=1, max_length=8)
    site_id: str | None = Field(default=None, max_length=36)
    registered_on: date | None = None
    excluded_on: date | None = None
    status: str | None = Field(default=None, max_length=16)
    responsible: str | None = Field(default=None, max_length=255)
    notes: str | None = None


class HazardousFacilityRead(BaseSchema):
    id: str
    name: str
    register_number: str
    hazard_class: str
    #: класс и состояние словами — перевод делает сервер, экран его не дублирует
    hazard_class_label: str
    site_id: str | None = None
    registered_on: date | None = None
    excluded_on: date | None = None
    status: str
    status_label: str
    responsible: str | None = None
    notes: str | None = None


class HazardousFacilityPage(BaseSchema):
    items: list[HazardousFacilityRead]
    total: int


class TechnicalDeviceCreate(BaseSchema):
    #: обязателен: экспертиза и надзор идут по зарегистрированному объекту,
    #: устройство «ничьё» невозможно предъявить проверяющему
    facility_id: str = Field(min_length=1, max_length=36)
    kind: str = Field(min_length=1, max_length=32)
    name: str = Field(min_length=1, max_length=255)
    serial_number: str | None = Field(default=None, max_length=64)
    commissioned_on: date | None = None
    lifetime_until: date | None = None
    epb_conclusion_number: str | None = Field(default=None, max_length=64)
    epb_registered_on: date | None = None
    epb_valid_until: date | None = None
    status: str = Field(default="in_operation", max_length=16)
    notes: str | None = None


class TechnicalDeviceUpdate(BaseSchema):
    facility_id: str | None = Field(default=None, min_length=1, max_length=36)
    kind: str | None = Field(default=None, min_length=1, max_length=32)
    name: str | None = Field(default=None, min_length=1, max_length=255)
    serial_number: str | None = Field(default=None, max_length=64)
    commissioned_on: date | None = None
    lifetime_until: date | None = None
    epb_conclusion_number: str | None = Field(default=None, max_length=64)
    epb_registered_on: date | None = None
    epb_valid_until: date | None = None
    status: str | None = Field(default=None, max_length=16)
    notes: str | None = None


class TechnicalDeviceRead(BaseSchema):
    id: str
    facility_id: str
    kind: str
    kind_label: str
    name: str
    serial_number: str | None = None
    commissioned_on: date | None = None
    lifetime_until: date | None = None
    epb_conclusion_number: str | None = None
    epb_registered_on: date | None = None
    epb_valid_until: date | None = None
    status: str
    status_label: str
    notes: str | None = None
    #: ok / due_soon / overdue / absent — считается ПРИ ЧТЕНИИ.
    #: «absent» — отдельное состояние, а не разновидность просрочки
    epb_status: str
    epb_status_label: str
    #: назначенный срок службы истёк — факт из данных, не суждение о том,
    #: обязана ли экспертиза быть проведена
    past_lifetime: bool = False
    #: разд. 54.2 «история работ»: последняя ПОДТВЕРЖДЁННАЯ работа — срок без
    #: неё это обещание, а не доказательство. Считается при чтении
    last_work_on: date | None = None
    last_work_result: str | None = None


class TechnicalDevicePage(BaseSchema):
    items: list[TechnicalDeviceRead]
    total: int


class DeviceWorkCreate(BaseSchema):
    device_id: str = Field(min_length=1, max_length=36)
    kind: str = Field(min_length=1, max_length=32)
    performed_on: date
    result: str = Field(min_length=1, max_length=32)
    performer: str | None = Field(default=None, max_length=255)
    #: обязателен для вида ``epb`` — см. валидацию в ручке
    conclusion_number: str | None = Field(default=None, max_length=64)
    notes: str | None = None
    next_due: date | None = None


class DeviceWorkRead(BaseSchema):
    id: str
    device_id: str
    kind: str
    kind_label: str
    performed_on: date
    result: str
    result_label: str
    performer: str | None = None
    conclusion_number: str | None = None
    notes: str | None = None
    next_due: date | None = None
    #: перенесла ли эта работа срок эксплуатации устройства; переносит только
    #: положительная ЭПБ
    shifted_due: bool = False


class DeviceWorkPage(BaseSchema):
    items: list[DeviceWorkRead]
    total: int


class PcPlanCreate(BaseSchema):
    year: int = Field(ge=2000, le=2100)
    title: str = Field(min_length=1, max_length=255)
    responsible: str | None = Field(default=None, max_length=255)
    approved_on: date | None = None
    status: str = Field(default="draft", max_length=16)
    notes: str | None = None


class PcPlanUpdate(BaseSchema):
    year: int | None = Field(default=None, ge=2000, le=2100)
    title: str | None = Field(default=None, min_length=1, max_length=255)
    responsible: str | None = Field(default=None, max_length=255)
    approved_on: date | None = None
    status: str | None = Field(default=None, max_length=16)
    notes: str | None = None


class PcPlanRead(BaseSchema):
    id: str
    year: int
    title: str
    responsible: str | None = None
    approved_on: date | None = None
    status: str
    status_label: str
    notes: str | None = None
    #: сколько мероприятий в плане и сколько из них просрочено — считается
    #: при чтении, чтобы план был виден одной строкой
    measures_total: int = 0
    measures_overdue: int = 0


class PcPlanPage(BaseSchema):
    items: list[PcPlanRead]
    total: int


class PcMeasureCreate(BaseSchema):
    plan_id: str = Field(min_length=1, max_length=36)
    section: str = Field(min_length=1, max_length=32)
    title: str = Field(min_length=1, max_length=255)
    due_on: date
    responsible: str | None = Field(default=None, max_length=255)
    status: str = Field(default="planned", max_length=16)
    completed_on: date | None = None
    result: str | None = None


class PcMeasureUpdate(BaseSchema):
    plan_id: str | None = Field(default=None, min_length=1, max_length=36)
    section: str | None = Field(default=None, min_length=1, max_length=32)
    title: str | None = Field(default=None, min_length=1, max_length=255)
    due_on: date | None = None
    responsible: str | None = Field(default=None, max_length=255)
    status: str | None = Field(default=None, max_length=16)
    completed_on: date | None = None
    result: str | None = None


class PcMeasureRead(BaseSchema):
    id: str
    plan_id: str
    section: str
    section_label: str
    title: str
    due_on: date
    responsible: str | None = None
    #: planned / overdue / done / cancelled — «просрочено» СЧИТАЕТСЯ при чтении
    status: str
    status_label: str
    completed_on: date | None = None
    result: str | None = None


class PcMeasurePage(BaseSchema):
    items: list[PcMeasureRead]
    total: int


class OpoAttestationRead(BaseSchema):
    """Аттестация по промбезопасности глазами дисциплины.

    Сама запись живёт в ЯДРЕ (``Attestation``) — дисциплина её не дублирует, а
    показывает СВОИ записи и добавляет то, чего у ядра нет: имя человека рядом
    и состояние срока словами.

    «Свои» — это записи с областью СВОЕЙ ДИСЦИПЛИНЫ. Здесь стояло «те, у
    которых заполнена область из справочника»: верно, пока справочник был
    перечнем Ростехнадзора, и ложно с разд. 56.2 срез-6, где в нём появилась
    область «ПДД» (проверка знаний водителей).
    """

    id: str
    person_id: str
    person_name: str
    name: str
    area_code: str
    area_label: str
    issued_at: date | None = None
    expires_at: date | None = None
    #: ok / due_soon / overdue / absent — считается ПРИ ЧТЕНИИ
    validity_status: str
    validity_status_label: str


class OpoAttestationPage(BaseSchema):
    items: list[OpoAttestationRead]
    total: int


class IndustrialReadinessRead(BaseSchema):
    """Сводка ПромБеза: сколько объектов и какого класса опасности.

    От класса зависит режим надзора (объекты I и II класса — постоянный
    государственный надзор и обязательная декларация промышленной
    безопасности), поэтому разрез по классам — не украшение, а первое, что
    нужно специалисту и проверяющему.

    Считаются ДЕЙСТВУЮЩИЕ объекты: исключённый из госреестра остаётся в
    системе ради истории, но объектом надзора быть перестаёт.
    """

    total_facilities: int
    #: класс → число действующих объектов; ключи — всегда все четыре, чтобы
    #: «ноль объектов I класса» отличался от «поле не пришло»
    by_class: dict[str, int]
    excluded_facilities: int
    #: разд. 54.2 «технические устройства… ЭПБ, сроки»: считаются только
    #: эксплуатируемые устройства — списанное просрочкой быть не может
    total_devices: int = 0
    epb_overdue: int = 0
    epb_due_soon: int = 0
    #: ФАКТ из данных: назначенный срок службы истёк, а действующего заключения
    #: ЭПБ нет. ГРАНИЦА: платформа НЕ решает, обязана ли экспертиза быть
    #: проведена — это зависит от типа устройства, документации и норм ФНП,
    #: которых в данных нет. Поэтому поля «устройств без ЭПБ» здесь НЕТ.
    devices_past_lifetime_without_epb: int = 0
    #: разд. 54.2 «история работ»: устройства, по которым нет НИ ОДНОЙ записи о
    #: работах — срок стоит, а подтвердить его нечем
    devices_without_work_record: int = 0
    #: разд. 54.2 «аттестация персонала»: считаются ТОЛЬКО записи с областью из
    #: справочника — сводка дисциплины показывает свои записи, а не все
    #: аттестации арендатора
    attestations_total: int = 0
    attestations_overdue: int = 0
    attestations_due_soon: int = 0
    #: разд. 54.2 «производственный контроль». ГРАНИЦА: платформа сообщает
    #: ФАКТ наличия плана на текущий год, но не объявляет его отсутствие
    #: нарушением — обязанность вести ПК зависит от того, эксплуатирует ли
    #: организация ОПО, и полноту сведений определяет специалист.
    current_year_plan_exists: bool = False
    pc_measures_overdue: int = 0
    pc_measures_planned: int = 0
